### PR TITLE
Fix EVP_Cipher() for provided cipher implementations

### PR DESCRIPTION
**Pull request by levitte**
EVP_Cipher() would return whatever ctx->cipher->ccipher() returned
with no regard for historical semantics.

We change this to first look if there is a ctx->cipher->ccipher(), and
in that case we treat the implementation as one with a custom cipher,
and "translate" it's return value like this: 0 => -1, 1 => outl, where
|outl| is the output length.

If there is no ctx->cipher->ccipher, we treat the implementation as
one without a custom cipher, call ctx->cipher->cupdate or
ctx->cipher->cfinal depending on input, and return whatever they
return (0 or 1).

Furthermore, we add a small hack in EVP_CIPHER_flags() to check if the
cipher is a provided one, and add EVP_CIPH_FLAG_CUSTOM_CIPHER to the
flags to be returned if there is a cipher->ccipher.  That way,
provided implementations never have to set that flag themselves, all
they need to do is to include a OSSL_FUNC_CIPHER_CIPHER function.

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by levitte** [comment]
This replaces #10125

**Comment by levitte** [comment]
I'm pretty sure that with this, we can clean away some of the blah_cipher() functions in some of the provided cipher implementations.  I'd say that's even desirable at this point, but when I've tried, it seems I'm doing too much or not doing it right.  So, er, @slontis?

File: doc/man3/EVP_EncryptInit.pod
Comment (mattcaswell) [inline doc/man3/EVP_EncryptInit.pod]: Are we going to deprecate this?

File: doc/man3/EVP_EncryptInit.pod
Comment (mattcaswell) [inline doc/man3/EVP_EncryptInit.pod]: I would simplify this. Just say that if `EVP_CIPH_FLAG_CUSTOM_CIPHER` is not set then you must only call this function with a multiple of the blocksize otherwise the behaviour is undefined and may lead to crashes.

File: doc/man3/EVP_EncryptInit.pod
Comment (mattcaswell) [inline doc/man3/EVP_EncryptInit.pod]: I'd expand this warning and recommend that applications don't use this at all and instead use the update/final.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: doc/man3/EVP_EncryptInit.pod
Comment (levitte) [reply doc/man3/EVP_EncryptInit.pod]: I think that should be left to another PR.  This fixes a blatant bug, let's stay with that.

File: doc/man3/EVP_EncryptInit.pod
Comment (levitte) [reply doc/man3/EVP_EncryptInit.pod]: Okie

**Comment by slontis** [comment]
>  I'd say that's even desirable at this point, but when I've tried, it seems I'm doing too much or not doing it right. So, er, @slontis?

Probably because the update and final call the cipher() internally for most cases..
So even if we remove the dispatch cipher function from the table the method is still useful internally..

I am thinking of maybe collapsing the einit/dinit into a single call also as most of the time the code ends up doing the logic to call a common function that passes the enc state anyway.

**Comment by levitte** [comment]
Re einit/dinit, I find the separated interface cleaner...

**Comment by slontis** [comment]
> Re einit/dinit, I find the separated interface cleaner...

I understand what you are saying, but all this really did was shift the same code into every cipher implementation. so instead of handling in 2 places in the high level code - we handle it everywhere we use it - so the logic you are making cleaner is not any cleaner in the lower level

**Comment by slontis** [comment]
oops didnt mean to close

**Comment by levitte** [comment]
Either way, einit/dinit should be debated elsewhere. Make an issue 😉

**Comment by levitte** [comment]
> Probably because the update and final call the cipher() internally for most cases..
So even if we remove the dispatch cipher function from the table the method is still useful internally..

What I meant to say was that if the CIPHER_CIPHER function essentially just points at an equivalent of the CIPHER_UPDATE function, there's no real reason to keep the CIPHER_CIPHER function around.
(BTW, I do think that I saw a file where the update and the cipher function were separate, but still exact duplicates of each other. I will have to look again)

**Comment by slontis** [comment]
Its not 1-1 currently because there is a difference between block and stream
see cipher_common.c

cipher_generic_block_update
cipher_generic_stream_update

cipher_generic_cipher

**Comment by levitte** [comment]
aes_xts_stream_update() is just a call of aes_xts_cipher(), so the latter can really just be collapsed in the former...

Same goes for tdes_wrap_update() and tdes_wrap_cipher()

**Comment by levitte** [comment]
Either way, please do note that I'm not doing any such collapsing in this PR.  Just saying aloud that one might want to pay attention and check if the CIPHER_CIPHER function actually needs being implemented (and yeah, I get the thing with cipher_generic_block_update, cipher_generic_stream_update and cipher_generic_cipher...  it took me a moment to wrap my brain, but I finally got there).  Food for thought for future PRs if anyone feels compelled

File: crypto/evp/evp_lib.c
Comment (slontis) [inline crypto/evp/evp_lib.c]: I dont know that this is a good idea at the moment... Maybe this PR should just deal with the return 0 issues..

Comment (levitte) [reply crypto/evp/evp_lib.c]: There's no real way around it unless you want to mess around with that flag in the implementations. After all, an existing ccipher that behaves like cupdate should be harmless, and because the returned value from EVP_Cipher is tied to that flag...

File: doc/man3/EVP_EncryptInit.pod
Comment (slontis) [inline doc/man3/EVP_EncryptInit.pod]: and leaves
OR <in> leaving

Comment (levitte) [reply doc/man3/EVP_EncryptInit.pod]: Fixed

**Comment by levitte** [comment]
Ping

File: doc/man3/EVP_EncryptInit.pod
Comment (mattcaswell) [inline doc/man3/EVP_EncryptInit.pod]: Better to reference the  full function names

**Review by mattcaswell — DISMISSED** [review]
Approved assuming the one comment is fixed.

File: doc/man3/EVP_EncryptInit.pod
Comment (levitte) [reply doc/man3/EVP_EncryptInit.pod]: Ugh...  it's a gazillion!

Comment (levitte) [reply doc/man3/EVP_EncryptInit.pod]: I restrained myself...  and I'll ask you to review the change.

Comment (richsalz) [reply doc/man3/EVP_EncryptInit.pod]: How about this:
```
This function is historic and shouldn't be used in an application.
Use the "update" and "final" functions (e.g., L<DESUpdate(3)>, L<DESFinal(3)>)
insead.
```

Comment (levitte) [reply doc/man3/EVP_EncryptInit.pod]: Why the DES functions?????

**Review by mattcaswell — APPROVED** [review]

File: doc/man3/EVP_EncryptInit.pod
Comment (richsalz) [reply doc/man3/EVP_EncryptInit.pod]: it was a sample.  pick whatever you want. :)

**Comment by levitte** [comment]
Thank you!

Merge.

f7397f0d58ce7ddf4c5366cd1846f16b341fbe43 Fix EVP_Cipher() for provided cipher implementations
ff33581c67928d2c21f1e853bca3a561335d4c3e Providers: fix OSSL_FUNC_CIPHER_CIPHER functions
5b084ca0492cee7aeca63e0a50dbff7487e3ee41 Remove EVP_CIPH_FLAG_CUSTOM_CIPHER in all our providers

**Pull request closed without merging**
